### DISCUSSION TITLE
How to use __subclasshook__ with Mypy?

### DISCUSSION THREAD

**Discussion by geryogam** [python/typing#1127] [Q&A] [answered]
How come that under Mypy, `__subclasshook__` works for one-trick ponies from `collections.abc`, but not for user-defined classes?

For instance, this program

```python
from collections.abc import Hashable

class A:
    def __hash__(self) -> int:
        return 0

a: Hashable = A()
```

outputs

```
$ mypy demo.py --strict
Success: no issues found in 1 source file
```

But this equivalent program

```python
from abc import ABCMeta, abstractmethod

def _check_methods(C: type, *methods: str) -> bool:
    mro = C.__mro__
    for method in methods:
        for B in mro:
            if method in B.__dict__:
                if B.__dict__[method] is None:
                    return NotImplemented
                break
        else:
            return NotImplemented
    return True

class Hashable(metaclass=ABCMeta):
    __slots__ = ()

    @abstractmethod
    def __hash__(self) -> int:
        return 0

    @classmethod
    def __subclasshook__(cls, C: type) -> bool:
        if cls is Hashable:
            return _check_methods(C, "__hash__")
        return NotImplemented

class A:
    def __hash__(self) -> int:
        return 0

a: Hashable = A()
```

outputs

```
$ mypy demo.py --strict
demo.py:32: error: Incompatible types in assignment (expression has type "A", variable has type "Hashable")
Found 1 error in 1 file (checked 1 source file)
```

Does Mypy handle one-trick ponies in a special way?

**Comment by erictraut** [comment]
`collections.abc.Hashable` is a Protocol class, conformant with PEP 544. If you make your custom `Hashable` class a Protocol class, it will work with static type checkers. The `__subclasshook__` and `register` methods are dynamic mechanisms that won't work with static type checking.

**Reply by geryogam** [reply]
Thanks for the reply @erictraut.

> `collections.abc.Hashable` is a Protocol class

Yet that is not what I see here (https://github.com/python/cpython/blob/v3.10.4/Lib/_collections_abc.py#L90):

```python
class Hashable(metaclass=ABCMeta):
```

**Reply by JelleZijlstra** [reply]
Type checkers don't look at the runtime implementation. What matters to mypy is the definition in typeshed: https://github.com/python/typeshed/blob/master/stdlib/_collections_abc.pyi#L16. That points to https://github.com/python/typeshed/blob/40553277ad7a7beccaa8698d72f3ecae9b1a81b8/stdlib/typing.pyi#L666 which is indeed a Protocol.

**Reply by geryogam** [reply]
Thanks @JelleZijlstra!

**Comment by AlexWaygood** [comment, marked as answer]
It's important to recognise that mypy has no knowledge of the runtime implementations of anything in the standard library (this is because most of the standard library doesn't have type annotations). The same goes for most other static type checkers.

For the standard library, all mypy knows about is the stubs that are provided for these classes in typeshed. You'll find the definition for `Hashable` in typeshed [here](https://github.com/python/typeshed/blob/master/stdlib/typing.pyi#L665), and you'll find that it's completely different to the implementation at runtime. This is why simply copying the runtime implementation produces code that mypy cannot understand — if you copy the implementation in typeshed instead, you should find that mypy is perfectly able to understand your code.

**Reply by geryogam** [reply]
Thanks!

**Reply by AlexWaygood** [reply]
No problem!
